File: 0005-longest-palindromic-substring/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n <= 1:
            return s
        dp = [[False] * n for _ in range(n)]
        longest = ""
        for x in range(n):
            dp[x][x] = True
            longest = s[0]
        for length in range(2, n + 1):
            for l in range(n - length + 1):
                r = l + length - 1
                if s[l] == s[r]:
                    if length == 2 or dp[l + 1][r - 1]:
                        dp[l][r] = True
                        if length > len(longest):
                            longest = s[l:r + 1]
        return longest

File: 0005-longest-palindromic-substring/solution.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        int longest = 0; int longestindex = 0;
        for(int i = n - 1; i >= 0; --i){
            for(int j = i; j < n; ++j){
                if(s[j] == s[i] and (j - i <= 2 or dp[i + 1][j - 1] == true)){
                    dp[i][j] = true;
                    if(j - i + 1 > longest){
                        longest = j - i + 1;
                        longestindex = i;
                    }
                }
            }
        }
        return s.substr(longestindex, longest);
    }
};
